examples/lighting-app/esp32/main/Button.cpp

```diff
@@ -15,52 +16,102 @@
+
+#include "AppTask.h"
 #include "Button.h"
-#include "esp_attr.h"
+#include "Globals.h"
+#include "ScreenManager.h"
+#include <lib/support/CodeUtils.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <vector>
 
-#define GPIO_INPUT_IO_0 9
-#define GPIO_INPUT_PIN_SEL (1ULL << GPIO_INPUT_IO_0)
-#define ESP_INTR_FLAG_DEFAULT 0
+static const char * TAG = "Button.cpp";
 
-static const char * TAG = "Button";
+extern Button gButtons[BUTTON_NUMBER];
 
-static Button::ButtonPressCallback button_press_handler = nullptr;
+Button::Button() {}
 
-static void IRAM_ATTR gpio_isr_handler(void * arg)
+Button::Button(gpio_num_t gpioNum)
 {
-    if (button_press_handler != nullptr)
+    mGPIONum = gpioNum;
+}
+
+int32_t Find_Button_Via_Pin(gpio_num_t gpioNum)
+{
+    for (int i = 0; i < BUTTON_NUMBER; i++)
     {
-        button_press_handler();
+        if (gButtons[i].GetGPIONum() == gpioNum)
+        {
+            return i;
+        }
     }
+    return -1;
 }
 
-void Button::Init()
+void IRAM_ATTR button_isr_handler(void * arg)
 {
-    /* Initialize button interrupt*/
-    // zero-initialize the config structure.
+    uint32_t gpio_num = (uint32_t) arg;
+    int32_t idx       = Find_Button_Via_Pin((gpio_num_t) gpio_num);
+    if (idx == -1)
+    {
+        return;
+    }
+    BaseType_t taskWoken = pdFALSE;
+    xTimerStartFromISR(gButtons[idx].mbuttonTimer,
+                       &taskWoken); // If the timer had already been started ,restart it will reset its expiry time
+}
+
+esp_err_t Button::Init()
+{
+    return Init(mGPIONum);
+}
+
+esp_err_t Button::Init(gpio_num_t gpioNum)
+{
+    esp_err_t ret = ESP_OK;
+
+    mGPIONum = gpioNum;
+    //  zero-initialize the config structure.
     gpio_config_t io_conf = {};
-    // interrupt of rising edge
+    // interrupt of falling edge
     io_conf.intr_type = GPIO_INTR_NEGEDGE;
     // bit mask of the pins, use GPIO4/5 here
-    io_conf.pin_bit_mask = GPIO_INPUT_PIN_SEL;
+    io_conf.pin_bit_mask = 1ULL << gpioNum;
     // set as input mode
     io_conf.mode = GPIO_MODE_INPUT;
     // enable pull-up mode
     io_conf.pull_up_en = GPIO_PULLUP_ENABLE;
+
     gpio_config(&io_conf);
 
-    // install gpio isr service
-    gpio_install_isr_service(ESP_INTR_FLAG_DEFAULT);
     // hook isr handler for specific gpio pin
-    gpio_isr_handler_add(static_cast<gpio_num_t>(GPIO_INPUT_IO_0), gpio_isr_handler, (void *) GPIO_INPUT_IO_0);
+    ret = gpio_isr_handler_add(gpioNum, button_isr_handler, (void *) gpioNum);
+    ESP_RETURN_ON_ERROR(ret, TAG, "gpio_isr_handler_add failed: %s", esp_err_to_name(ret));
 
-    ESP_LOGI(TAG, "Button initialized..");
-}
+    mbuttonTimer = xTimerCreate("BtnTmr",               // Just a text name, not used by the RTOS kernel
+                                pdMS_TO_TICKS(50),      // timer period
+                                false,                  // no timer reload (==one-shot)
+                                (void *) (int) gpioNum, // init timer id = gpioNum index
```

Comment:
```suggestion
                                (void *) (uint8_t) gpioNum, // init timer id = gpioNum index
```

Reply:
I believe you are correct, however I am afraid to miss something here - this was copied over from existing Button.cpp so I would prefer to keep it to have the files somewhat in sync for when we want to deduplicate them.